tools/2024-09-gitops-flux.md

```diff
@@ -0,0 +1,289 @@
+
+#### Bran can use GitOps to deploy and manage a Radius application in production
+
+Bran is a developer who uses GitOps to automate the deployment of Radius applications to production Kubernetes clusters. His infrastructure team has set up a Radius installation and environment on the production cluster, and Bran wants to deploy his application using Radius. He has authored a Bicep manifest that defines the resources for his application and a Bicep parameters file that specifies the parameters for the manifest. He commits the Bicep manifest and parameters file to a Git repository.
+
+When the GitOps tool watches the repository and applies the Bicep manifest and parameters file to the production cluster, the Radius resources defined in the Bicep manifest are deployed. The application is deployed and Bran can manage the application using GitOps practices.
+
+#### Sam can use GitOps to patch a Radius application in production
+
+Sam is an SRE who is responsible for maintaining a Radius application in production. He notices that the application is experiencing performance issues and the number of application container replicas needs to be increased. He updates the Bicep parameters file for the application to specify the new number of replicas, and commits the changes to a Git repository.
+
+When the GitOps tool watches the repository and applies the BicepDeployment resource to the production cluster, the Radius resources defined in the Bicep manifest are updated with the new parameters. The application container replicas are increased, and the performance issues are resolved.
+
+## Design
+
+### High Level Design
+<!--
+High level overview of the data flow and key components.
+
+Provide a high-level description, using diagrams as appropriate, and top-level
+explanations to convey the architectural/design overview. Don’t go into a lot
+of details yet but provide enough information about the relationship between
+these components and other components. Call out or highlight new components
+that are not part of this feature (dependencies). This diagram generally
+treats the components as black boxes. Provide a pointer to a more detailed
+design document, if one exists. 
+-->
+
+This design proposes a new component, the Flux Controller, that integrates Radius with Flux. It will be implemented as a Kubernetes Controller following the [flux source watcher](https://fluxcd.io/flux/gitops-toolkit/source-watcher/) extensibility pattern. This controller will be responsible for reacting to changes to `.bicep` and `.bicepparam` files in the Flux source controller, compiling the Bicep manifests, and creating or updating `BicepDeployment` resources on the Kubernetes cluster. These `BicepDeployment` resources will in turn be managed by the BicepDeployment Controller, which will be responsible for interacting with Radius directly to deploy and manage these resources.
```

Comment:
curious: are there limitations to what all can be configured using bicepparams? if so, do we need to call it out?